
**[DSLC stage 2]: Data cleaning and pre-processing**


## Domain problem formulation

*Our goal is  to identify countries that have demonstrated an increase in organ transplant donation rates over time, and which countries have the highest organ donation rates.*


## Data source overview

This data comes from the publicly available survey data from the Global Observatory on Donation and Transplantation (GODT) that was collected in a collaboration between World Health Organization (WHO) and the Spanish Transplant Organization, Organización Nacional de Trasplantes (ONT). The data portal can be found at http://www.transplant-observatory.org/export-database/.

This database contains information about organ donations and transplants (a total of 24 variables) for 194 countries and is collected every year based on a survey that, according to the website, began in 2007. We will consider a version of the data that contains information up to 2017 (downloaded in 2018) for this project.



In [1]:
import pandas as pd
import numpy as np
import plotly.express as px

pd.set_option('display.max_columns', None)
pd.options.mode.copy_on_write = True

The data is contained within the file `global-organ-donation_2018.csv`.


## Step 1: Review background information


- *What does each variable measure?* A data dictionary is presented in the subsection below.

- *How was the data collected?* The website ([http://www.transplant-observatory.org/methodology/](http://www.transplant-observatory.org/methodology/)) from which we downloaded the organ donation data states that the data is based on a survey that began in 2007, and is sent annually via email to "national focal points".

- *What are the observational units?* To identify the observational units, consider for what entities each full set of organ donation measurements is collected. For the organ donation data, these are the year and country combinations (a complete set of measurements are taken for each country, every year), which we will call the "country-years".

- *Is the data relevant to my project?* Yes it is a comprehensive public global summary of organ donations spanning several years.

- *What questions do I have and what assumptions am I making?*



### Data dictionary

The data dictionary we found on the website at the time of data collection is printed below:



From the GODT resources (https://tts.org/images/GODT/2020-Global-report-para-web-1.pdf):

- "Actual deceased donor": Deceased person from whom at least one organ has been recovered for the purpose of transplantation.

- "Utilized deceased donor": An actual donor from whom at least one organ has been transplanted.



> **Question: What proportion of deceased donors are "utilized" deceased donors?**
>
> Do most organs recovered for transplantation actually get transplanted?



## Step 2: Loading in the data


Let's load in the data.


In [ ]:
# If you upload the file locally to colab
# organs_original = pd.read_csv("global-organ-donation_2018.csv")

In [2]:
# Using the URL for the file
url = 'https://drive.google.com/file/d/1T-AsACpKRiFQElM1PKkENLQPhSO0n4ew/view?usp=sharing'
file_id = url.split('/')[-2]
dwn_url = 'https://drive.google.com/uc?id=' + file_id
organs_original = pd.read_csv(dwn_url)

The data column names match the names in the data dictionary.

In [3]:
#TODO: write a line of code to list the columns or variable names of the data
organs_original.columns

Index(['REGION', 'COUNTRY', 'REPORTYEAR', 'POPULATION', 'TOTAL Actual DD',
       'Actual DBD', 'Actual DCD', 'Total Utilized DD', 'Utilized DBD',
       'Utilized DCD', 'DD Kidney Tx', 'LD Kidney Tx', 'TOTAL Kidney Tx',
       'DD Liver Tx', 'DOMINO Liver Tx', 'LD Liver Tx', 'TOTAL Liver TX',
       'Total Heart', 'DD Lung Tx', 'LD Lung Tx', 'TOTAL Lung Tx',
       'Pancreas Tx', 'Kidney Pancreas Tx', 'Small Bowel Tx'],
      dtype='object')


Let's look at the first 20 rows of the dataset. Note that *all of the values after the first 4 identifier columns are missing for these first 20 rows*. We checked the data manually to make sure that this was not a data loading error, and it does seem that the data has been loaded in correctly.


In [6]:
#TODO: write a line of code to show the first 20 rows
organs_original.head(20)

REGION                 COUNTRY  REPORTYEAR  POPULATION  \
0                  Europe                 Andorra        2000         0.1   
1   Eastern Mediterranean    United Arab Emirates        2000         2.4   
2   Eastern Mediterranean             Afghanistan        2000        22.7   
3                 America     Antigua and Barbuda        2000         0.1   
4                  Europe                 Albania        2000         3.1   
5                  Europe                 Armenia        2000         3.5   
6                  Africa                  Angola        2000        12.9   
7                 America               Argentina        2000        37.0   
8                  Europe                 Austria        2000         8.2   
9         Western Pacific               Australia        2000        18.9   
10                 Europe              Azerbaijan        2000         7.7   
11                 Europe  Bosnia and Herzegovina        2000         4.0   
12                America                Barbados        2000         0.3   
13        South-East Asia              Bangladesh        2000       129.2   
14                 Europe                 Belgium        2000        10.2   
15                 Africa            Burkina Faso        2000        11.9   
16                 Europe                Bulgaria        2000         8.2   
17  Eastern Mediterranean                 Bahrain        2000         0.6   
18                 Africa                 Burundi        2000         6.7   
19                 Africa                   Benin        2000         6.1   

    TOTAL Actual DD  Actual DBD  Actual DCD  Total Utilized DD  Utilized DBD  \
0               NaN         NaN         NaN                NaN           NaN   
1               NaN         NaN         NaN                NaN           NaN   
2               NaN         NaN         NaN                NaN           NaN   
3               NaN         NaN         NaN                NaN           NaN   
4               NaN         NaN         NaN                NaN           NaN   
5               NaN         NaN         NaN                NaN           NaN   
6               NaN         NaN         NaN                NaN           NaN   
7               NaN         NaN         NaN                NaN           NaN   
8             194.0       194.0         0.0                NaN           NaN   
9             195.0       195.0         NaN                NaN           NaN   
10              NaN         NaN         NaN                NaN           NaN   
11              NaN         NaN         NaN                NaN           NaN   
12              NaN         NaN         NaN                NaN           NaN   
13              NaN         NaN         NaN                NaN           NaN   
14            256.0       256.0         NaN                NaN           NaN   
15              NaN         NaN         NaN                NaN           NaN   
16             26.0        26.0         NaN                NaN           NaN   
17              NaN         NaN         NaN                NaN           NaN   
18              NaN         NaN         NaN                NaN           NaN   
19              NaN         NaN         NaN                NaN           NaN   

    Utilized DCD  DD Kidney Tx  LD Kidney Tx  TOTAL Kidney Tx  DD Liver Tx  \
0            NaN           NaN           NaN              NaN          NaN   
1            NaN           NaN           NaN              NaN          NaN   
2            NaN           NaN           NaN              NaN          NaN   
3            NaN           NaN           NaN              NaN          NaN   
4            NaN           NaN           NaN              NaN          NaN   
5            NaN           NaN           NaN              NaN          NaN   
6            NaN           NaN           NaN              NaN          NaN   
7            NaN           NaN           NaN              NaN          NaN   
8            NaN  

We also want to print a *random* sample of 20 rows from the data below.

In [ ]:
#TODO: write a line of code to show a random sample of 20 rows

Next, we check the dimension of the data.

In [ ]:
#TODO: write a line of code to check the dimension of the data
# Recall: the dimension is the number of rows and columns

When we read in data we do *sanity checks* to make sure we loaded the data correctly.

**Sanity check:**  Is the number of rows divisible by the number of countries? The number of countries in the data is:

In [ ]:
#TODO: Write a line of code to figure out the number of countries

> **Question: What to do if the number of rows in the data does not match what we expect?**
>
>
First, calculate what you would expect the number of rows to be. Then we can explore the data to see what may be wrong...

**Let's reconvene class now to figure out how..**





## Step 3: Examine the data and create action items


In this section, we will look at the data itself to try to identify any invalid values, understand the missing values, and any abnormalities in the data.




### Finding invalid values



Below, we print out a summary of the values of each numeric column.


In [ ]:
organs_original.select_dtypes('number').describe()

REPORTYEAR   POPULATION  TOTAL Actual DD   Actual DBD   Actual DCD  \
count  3165.000000  3165.000000      1290.000000  1181.000000   959.000000   
mean   2007.710269    35.237283       310.726357   289.475868    38.872784   
std       4.793700   131.945656      1002.027550   922.151429   197.115786   
min    2000.000000     0.000000         0.000000     0.000000     0.000000   
25%    2004.000000     1.900000         5.000000     4.000000     0.000000   
50%    2008.000000     7.400000        58.000000    53.000000     0.000000   
75%    2012.000000    23.600000       194.750000   182.000000     2.000000   
max    2017.000000  1393.800000     10286.000000  8403.000000  3298.000000   

       Total Utilized DD  Utilized DBD  Utilized DCD  DD Kidney Tx  \
count         492.000000    514.000000    525.000000   1308.000000   
mean          285.699187    248.227626     39.636190    487.107034   
std           972.149942    809.979679    223.237394   1430.753024   
min             0.000000      0.000000      0.000000      0.000000   
25%             0.000000      1.000000      0.000000      8.000000   
50%            28.000000     33.500000      0.000000     95.500000   
75%           161.000000    150.750000      0.000000    346.000000   
max          9706.000000   8075.000000   3243.000000  14827.000000   

       LD Kidney Tx  TOTAL Kidney Tx  DD Liver Tx  DOMINO Liver Tx  \
count   1370.000000      1401.000000  1217.000000       885.000000   
mean     303.140876       752.231263   243.136401         1.447458   
std      829.967985      2049.562922   772.811183         5.755855   
min        0.000000         0.000000     0.000000         0.000000   
25%       12.250000        43.000000     0.000000         0.000000   
50%       51.000000       163.000000    30.000000         0.000000   
75%      190.000000       494.000000   138.000000         0.000000   
max     6647.000000     20638.000000  7715.000000        71.000000   

       LD Liver Tx  TOTAL Liver TX  Total Heart   DD Lung Tx  LD Lung Tx  \
count  1100.000000     1253.000000  1139.000000   592.000000  581.000000   
mean     45.695455      277.059856    78.051800    56.121622    0.203098   
std     140.703985      807.568306   283.838158   232.324876    1.710028   
min       0.000000        0.000000     0.000000     0.000000    0.000000   
25%       0.000000        2.000000     0.000000     0.000000    0.000000   
50%       2.000000       35.000000    11.000000     0.000000    0.000000   
75%      18.000000      178.000000    53.500000    30.000000    0.000000   
max    1200.000000     8082.000000  3273.000000  2478.000000   20.000000   

       TOTAL Lung Tx  Pancreas Tx  Kidney Pancreas Tx  Small Bowel Tx  
count    1017.000000   921.000000         1053.000000      910.000000  
mean       59.720747    41.018458           29.842355        3.720879  
std       216.100281   163.899852          111.359410       20.060555  
min         0.000000     0.000000            0.000000        0.000000  
25%         0.000000     0.000000            0.000000        0.000000  
50%         3.000000     3.000000            2.000000        0.000000  
75%        36.000000    24.000000           17.000000        0.000000  
max      2478.000000  1484.000000          924.000000      198.000000

There don't appear to be any negative count values, but the population seems to be on a strange scale (why are there countries whose population is 0, why is the largest population just 1393.8, and how can you have .8 of a person?).

The table below shows the average recorded population for a sample of 20 countries.


In [ ]:
organs_original.groupby("COUNTRY")["POPULATION"] \
    .mean() \
    .sample(20)

COUNTRY
Guyana                                         0.780000
Zimbabwe                                      13.060000
Grenada                                        0.100000
Bosnia and Herzegovina                         3.911765
The former Yugoslav Republic of Macedonia      2.055556
Gambia                                         1.613333
Argentina                                     40.250000
Yemen                                         22.233333
Comoros                                        0.766667
Slovakia                                       5.422222
Uzbekistan                                    27.243750
Madagascar                                    19.426667
Monaco                                         0.000000
Tunisia                                       10.312500
Belgium                                       10.688889
Niger                                         14.406667
Venezuela (Bolivarian Republic of)            27.800000
Costa Rica                              


How can we fix the populations?

TODO: In the following text cell, create an action item to fix this!

> **Data cleaning action item to fix population.**

TODO: add action item here.




Taking another look at the first 20 rows printed in the table above (in the data loading section), we also notice that the year pre-dates 2007, which was when the survey supposedly began.


> **Question: Why does the data contain years pre-2007**
>
> The data contains information prior to 2007, which was when the data collection survey supposedly began. Why is this the case? We couldn't find information online to answer this question, but perhaps this is due to back-reporting (i.e. countries providing historical data), or perhaps it is a mistake in the documentation.


> **Data cleaning action item: Add an option to remove the pre-2007 data**
>
> We want to have an option in our cleaning function to remove the pre-2007 data, but the default option will be to keep it.


To investigate if there are any strange values in the `TOTAL Actual DD` variable, the display below a histogram of the `TOTAL Actual DD` variable.

In [ ]:
#copy previous data frame
former_organs_original = organs_original
#TODO: add a line of code to remove all records before 2007
# We could add this to cleaning function, but will do here


In [ ]:
#TODO: Create histogram of TOTAL Actual DD values

**Example Sanity Checks**

The distribution should heavily skewed with a lot of 0s.



**Annotate each cell to explain what the sanity checks are doing. Explain why the results are what you expect or not**


In [ ]:
sum(organs_original["TOTAL Actual DD"] > organs_original["POPULATION"] * 1000000)


0

In [ ]:
sum(organs_original["TOTAL Actual DD"] < organs_original["Actual DCD"])

0

In [ ]:
sum(organs_original["TOTAL Actual DD"] < organs_original["Actual DBD"])

0

In [ ]:
sum(organs_original["TOTAL Actual DD"] < organs_original["Total Utilized DD"])

0

In [ ]:
sum(organs_original["TOTAL Actual DD"] < organs_original["TOTAL Kidney Tx"])

897

In [ ]:
sum(organs_original["TOTAL Actual DD"] < organs_original["TOTAL Liver TX"])

141

In [ ]:
sum(organs_original["TOTAL Actual DD"] < organs_original["Total Heart"])

0

### [problem-specific] Checking COUNTRY and YEAR combinations are unique

It occurs to us to check that each country-year combination is unique (i.e., there are no duplicated entries in the data). The table below counts the number of times each country-year combination appears in the data and shows the distinct counts. Since the only entry is 1, this implies that each country-year combination only appears once in the data, which is good.

In [ ]:
#TODO: Write a line of code to verify that each row in the table is unique

### Examining missing values

From our tables and histograms above, it doesn't look like there are any extreme values (such as `999`) that are hiding missing values. To make sure, we will plot a histogram of a few of the variables.

The output below shows the number and proportion of missing rows in each column, arranged in order of least to most missingness.



In [ ]:
missing_prop = organs_original.isna().sum() / len(organs_original.index)
missing_prop.sort_values()

Other than the descriptor and ID variables of `REGION`, `COUNTRY`, `REPORTYEAR`, and `POPULATION`, there are a LOT of missing values in this dataset. Almost 60% of the `TOTAL Actual DD` values (our variable of interest) are missing!

The figure below shows the distribution of missing data across the entire dataset.


In [ ]:
px.imshow(organs_original.isna().astype(int),
          color_continuous_scale='Greys')

*italicized text*
How this missingness is distributed over time? The figure below shows the number of *non-missing* values reported for each year.

In [ ]:
missing_values_by_year = (organs_original.set_index("REPORTYEAR")["TOTAL Actual DD"] # set the index to be REPORTYEAR and select just TOTAL Actual DD col
                                         .notna() # identify whether each TOTAL Actual DD value is not missing
                                         .groupby(level=0) # group by REPORTYEAR (index)
                                         .sum()) # add up the total number of missing TOTAL Actual DD values each year
missing_values_by_year

px.bar(missing_values_by_year, labels={"value": "Number of non-missing TOTAL Actual DD values"})

What about the distribution of missingness by country? Is it the case that there are countries for which individual years of data are missing? Or is it instead that all of the data are missing or none of it is?

The table below shows the number of non-missing `TOTAL Actual DD` entries for a random sample of 20 countries. There seems to be a really big range of missingness patterns across the countries! Some countries report literally no non-missing data, whereas others report data for 7, 10, or 13 years.


In [ ]:

# identify whether each TOTAL Actual DD value is not missing
# and group by COUNTRY (index)
# and add up the number of non-missing values
non_missing_by_country = organs_original.set_index("COUNTRY")["TOTAL Actual DD"] \
    .notna() \
    .groupby(level=0) \
    .sum()

# sample 20 random countries (with a seed)
# and arrange countries alphabetically
non_missing_by_country.sample(20, random_state=1303) \
    .sort_index()


COUNTRY
Algeria                              7
Antigua and Barbuda                  0
Austria                             11
Bahamas                              0
Bolivia (Plurinational State of)     8
Comoros                              0
Cook Islands                         0
Dominican Republic                  11
Iran (Islamic Republic of)          10
Israel                              11
Latvia                              11
Malawi                               0
Myanmar                              7
Pakistan                             7
Portugal                            11
Samoa                                0
Sierra Leone                         0
Sudan                               10
Sweden                              11
Togo                                 0
Name: TOTAL Actual DD, dtype: int64

Let's visualize the distribution of non-missing data by country in the figure below

In [ ]:
px.histogram(non_missing_by_country,
             nbins=19,
             labels={"value": "Number of non-missing TOTAL Actual DD values per country"})

What can we learn from the missingness? Do more countries report data or not?



Austria reports data every year. What query can we use to see what Austria reports for **TOTAL Actual DD**?


In [ ]:
# filter to just austria
organs_austria = ...
# print out just the country, year, and total actual DD columns



Does Peru report data for every year greater than 2007 in the study?

In [ ]:
# filter to just peru
organs_peru = ...
# print out just the country, year, and total actual DD columns


COUNTRY  REPORTYEAR  TOTAL Actual DD
1495    Peru        2007              NaN
1689    Peru        2008              NaN
1883    Peru        2009              NaN
2077    Peru        2010             94.0
2271    Peru        2011            127.0
2465    Peru        2012             94.0
2659    Peru        2013             97.0
2853    Peru        2014             73.0
2991    Peru        2015             82.0
3080    Peru        2016             70.0
3151    Peru        2017             52.0

To handle missing values we use **imputation** depending on the analysis that we want to conduct.

Some reasonable action items for dealing with the missingness might be to replace each missing count for each country with:

- The *average* of the two surrounding non-missing values from the country.

- The *closest* (in terms of year) non-missing value from the country. If the missing value is equidistant between two non-missing years, we could choose one at random.

- The *previous* non-missing value from the country.

- An *interpolated* value that takes into account the trend from all of the non-missing values from the country.

For the countries that do not report *any* donor counts (i.e., all their data is missing), we will impute their donor counts with 0. Here we are making an *assumption* that the countries that have entirely missing data do not have an organ donation system in place. This is probably not an entirely realistic assumption, however since there may be some countries that do have an organ donor system but chose not to report any data to the GODT, but since this data does not exist in the public domain, there is nothing else that we can really do about it.



> **Pre-processing action item: Impute the donor count variable**
>
> There are several judgment call options that seem reasonable for creating imputed donor count variables, including:
>
> - The *average* of the two surrounding non-missing values from the country.
>
> - The *previous* non-missing value from the country.
>
> For the countries that do not report *any* donor counts (i.e., all their data is missing), we will impute their donor counts with 0.


Other options include imputing using the *closest* (in terms of year) non-missing value from the country and using an *interpolated* value that takes into account the trend from all of the non-missing values from the country, but these are much trickier to code so we exclude them here.



#### Imputation function

Below, we define a function, `impute_feature()`, that we will later call inside our data cleaning/pre-processing function to impute the missing values, with arguments for each imputation judgment call option.


In [ ]:
def impute_feature(data, feature, group, impute_method="average"):
    impute_method = impute_method.lower()

    if impute_method == "average":
        # create two temporary variables each equal to feature
        data = data.assign(imputed_feature_tmp_prev=data[feature], imputed_feature_tmp_next=data[feature])
        # fill the first variable using forward fill
        data["imputed_feature_tmp_prev"] = data.groupby(group)["imputed_feature_tmp_prev"].ffill()
        # fill the second variable using backward fill
        data["imputed_feature_tmp_next"] = data.groupby(group)["imputed_feature_tmp_next"].bfill()
        # then define feature_imputed column to be the mean of the forward and backward filled values
        data['feature_imputed'] = data[['imputed_feature_tmp_next', 'imputed_feature_tmp_prev']].mean(axis=1, skipna=True)
        # impute any remaining missing values with 0
        data['feature_imputed'] = data['feature_imputed'].fillna(0)
        # remove the two temporary variables
        data = data.drop(columns=['imputed_feature_tmp_prev', 'imputed_feature_tmp_next'])
        return data["feature_imputed"]
    else:
        raise ValueError

We can compare the first 20 original and imputed TOTAL Actual DD donor counts using:


In [ ]:
organs_original["imputed_donors"] = impute_feature(organs_original,
                                                   feature = "TOTAL Actual DD",
                                                   group = "COUNTRY",
                                                   impute_method = "average")
organs_original[["COUNTRY", "REPORTYEAR", "TOTAL Actual DD", "imputed_donors"]].sample(20, random_state=1010)


### Assessing Column names


The column names in this dataset are a mess! Let's add an action item to clean them so that they are tidily formatted and human-readable.

> **Data cleaning action item: Clean the column names**
>
> Rename the columns so that they are consistently formatted, with underscore-separated words and human readable. Since we want to change the names of the variables themselves, we will do this manually.


The table below displays the column name conversion:

| New variable name | Original variable name |
| :---        |    :----   |
| `region` | `REGION` |
| `country` | `COUNTRY` |
| `year` | `REPORTYEAR` |
| `population` | `POPULATION` |
| `total_deceased_donors` | `TOTAL Actual DD` |
| `deceased_donors_brain_death` | `Actual DBD` |
| `deceased_donors_circulatory_death` | `Actual DCD` |
| `total_utilized_deceased_donors` | `Total Utilized DD` |
| `utilized_deceased_donors_brain_death` | `Utilized DBD` |
| `utilized_deceased_donors_circulatory_death` | `Utilized DCD` |
| `deceased_kidney_tx` | `DD Kidney Tx` |
| `living_kidney_tx` | `LD Kidney Tx` |
| `total_kidney_tx` | `TOTAL Kidney Tx` |
| `deceased_liver_tx` | `DD Liver Tx` |
| `living_liver_tx` | `LD Liver Tx` |
| `domino_liver_tx` | `DOMINO Liver Tx` |
| `total_liver_tx` | `TOTAL Liver TX` |
| `total_heart_tx` | `Total Heart` |
| `deceased_lung_tx` | `DD Lung Tx` |
| `living_lung_tx` | `DD Lung Tx` |
| `total_lung_tx` | `TOTAL Lung Tx` |
| `total_pancreas_tx` | `Pancreas Tx` |
| `total_kidney_pancreas_tx` | `Kidney Pancreas Tx` |
| `total_small_bowel_tx` | `Small Bowel Tx` |


We will officially only change the column names when we actually clean our data below, so the remaining explorations until then will still use the original column names.




### Assessing variable type

The table below prints out the class/type of each column in the data.


In [ ]:
organs_original.dtypes

REGION                 object
COUNTRY                object
REPORTYEAR              int64
POPULATION            float64
TOTAL Actual DD       float64
Actual DBD            float64
Actual DCD            float64
Total Utilized DD     float64
Utilized DBD          float64
Utilized DCD          float64
DD Kidney Tx          float64
LD Kidney Tx          float64
TOTAL Kidney Tx       float64
DD Liver Tx           float64
DOMINO Liver Tx       float64
LD Liver Tx           float64
TOTAL Liver TX        float64
Total Heart           float64
DD Lung Tx            float64
LD Lung Tx            float64
TOTAL Lung Tx         float64
Pancreas Tx           float64
Kidney Pancreas Tx    float64
Small Bowel Tx        float64
imputed_donors        float64
dtype: object

### [Exercise: to complete] Examining data completeness

While we checked that the ID variables (country and year combination) were *unique*, we didn't check whether all of the ID variables appear in the data (i.e., whether the data is *complete*). One way to check this is to confirm that every country has 18 rows in the data.

The table below shows the number of rows for each 20 randomly chosen countries. There are clearly NOT 18 rows for every country! In fact, there are almost never 18 rows for every country.

In [ ]:
organs_original.value_counts("COUNTRY").sample(20, random_state=383)

COUNTRY
Lebanon         16
Kiribati        15
Zambia          15
Saudi Arabia    18
Cook Islands    15
Indonesia       15
Uzbekistan      16
Cyprus          18
Saint Lucia     15
Samoa           15
Sierra Leone    15
Serbia          16
Bahamas         15
Honduras        18
Iceland         18
Malawi          15
Netherlands     18
Armenia         18
India           16
Burundi         15
Name: count, dtype: int64

This is clearly related to our earlier question of "Why does the number of rows in the data not match what we expect?". It is left as an exercise to see if you can figure out what is going on. We recommend starting by comparing the number of non-missing `TOTAL Actual DD` values with the number of rows each year. Does something change around 2015?

To get you started, here is a bar chart counting the number of rows in the data for each year.

In [ ]:
px.histogram(organs_original, x="REPORTYEAR")





**Exploration idea:**
Modify the code above to overlay another bar chart (using a different color) of the number of non-missing `TOTAL Actual DD` counts.


> **Data cleaning action item: Complete the data**
>
> Complete the data by adding in the missing rows and filling them with NA values. This should be done *prior* to missing value imputation so that the missing values we create are filled in during imputation.



Having concluded the suggested explorations, let's now address any remaining questions and assumptions that we have.






### [problem-specific] Checking the hierarchy of donor count variables

Recall that we assumed that the `TOTAL Actual DD` donor count is made up of the sum of the `Actual DBD` (brain-death) and `Actual DCD` (circulatory death) donor counts. Let's check to see whether this is actually the case.

If we ignore cases where there are missing values in any one of the three variables, then 99.3% of the remaining 945 rows satisfy the  `TOTAL Actual DD` = `Actual DBD` + `Actual DCD`:


In [ ]:
# drop rows with missing values in any of the three columns
organs_no_missing = organs_original.dropna(subset=["TOTAL Actual DD", "Actual DBD", "Actual DCD"])
# compute the proportion of rows for which TOTAL Actual DD is equal to Actual DBD + Actual DCD
sum(organs_no_missing["TOTAL Actual DD"] == organs_no_missing["Actual DBD"] + organs_no_missing["Actual DCD"]) / len(organs_no_missing.index)

0.9925925925925926

(However, note that only 30% of the rows have non-missing values in all three variables.)



## Step 4: Clean and pre-process the data

Now we're ready to actually clean the data. Throughout our explorations above, we created the following action items (listed in the order they will be implemented):

- Rename the column names so that they are consistent, human-readable, underscore-separated, and lowercase

- Complete the data by adding in absent rows. The entries will be filled with missing values.

- Multiply the population variable by 1 million

We also created the following pre-processing action items (that are not necessary for the data to be clean, but they will be useful for our analyses):

- Add an imputed version of the `TOTAL Actual DD` count variable. There are several judgment call options for this step ("average", "previous", "closest", "interpolate"). We could also add imputed versions of the other variables too^[As an exercise, you may want to try and modify the `impute_feature()` code to also impute some of the transplant variables.], but we will focus on this variable in our analysis, we opted to just impute this variable (as well as the population variable values that were filled as missing when we completed the data).

- Add an option to remove (or not) the pre-2007 data.

To keep things simple, rather than writing a separate pre-processing function, we just wrote a single "data preparation" function that included both the cleaning and pre-processing action items:

We saved the cleaning function, `prepare_organ_data()` in the file `functions/prepare_organ_data.py`. This function makes use of the `impute_feature()` imputation function that we wrote which can be found in the `functions/impute_feature.py` file.  The `prepare_organ_data()` function is reproduced below.


In [ ]:
import pandas as pd
#from functions.impute_feature import impute_feature


def prepare_organ_data(organs_original,
                       impute_method = "average",
                       per_mil_vars = True):


  # define a cleaned version of the original organs data
  # rename the original rows
  organs_clean = organs_original.rename(columns={
    'REGION': 'region',
    'COUNTRY': 'country',
    'REPORTYEAR': 'year',
    'POPULATION': 'population',
    'TOTAL Actual DD': 'total_deceased_donors',
    'Actual DBD': 'deceased_donors_brain_death',
    'Actual DCD': 'deceased_donors_circulatory_death',
    'Total Utilized DD': 'total_utilized_deceased_donors',
    'Utilized DBD': 'utilized_deceased_donors_brain_death',
    'Utilized DCD': 'utilized_deceased_donors_circulatory_death',
    'DD Kidney Tx': 'deceased_kidney_tx',
    'LD Kidney Tx': 'living_kidney_tx',
    'TOTAL Kidney Tx': 'total_kidney_tx',
    'DD Liver Tx': 'deceased_liver_tx',
    'LD Liver Tx': 'living_liver_tx',
    'DOMINO Liver Tx': 'domino_liver_tx',
    'TOTAL Liver TX': 'total_liver_tx',
    'Total Heart': 'total_heart_tx',
    'DD Lung Tx': 'deceased_lung_tx',
    'DD Lung Tx': 'living_lung_tx',
    'TOTAL Lung Tx': 'total_lung_tx',
    'Pancreas Tx': 'total_pancreas_tx',
    'Kidney Pancreas Tx': 'total_kidney_pancreas_tx',
    'Small Bowel Tx': 'total_small_bowel_tx'}).copy()

  # add the rows with missing country-year combinations
  country_year_combinations = pd.MultiIndex.from_product([organs_clean[col].unique() for col in ["country", "year"]],
                                                         names=["country", "year"])
  # put these combinations into a data frame
  country_year_combinations_df = pd.DataFrame(index=country_year_combinations).reset_index()
  organs_clean = country_year_combinations_df.merge(organs_clean, on=["country", "year"], how="left")

  # For newly added rows, fill region with the unique values from the pre-existing rows
  organs_clean["region"] = organs_clean.groupby("country")["region"].transform(lambda x: x.ffill().bfill())

  # multiply the population variable by 1 million
  organs_clean["population"] = organs_clean["population"] * 1000000

  # add imputed features using the specified imputation method
  # impute_feature() is a custom function defined in imputeFeature.R
  # Note that we are only imputing the total_deceased_donors variable and the
  # population variable (missing values were introduced when we
  # "completed" the data). You could impute more variables if you wanted to.
  if impute_method in ["average", "previous"]:
    organs_clean["population_imputed"] = impute_feature(organs_clean,
                                                        feature = "population",
                                                        group = "country",
                                                        impute_method = impute_method)
    organs_clean["total_deceased_donors_imputed"] = impute_feature(organs_clean,
                                                                  feature = "total_deceased_donors",
                                                                  group = "country",
                                                                  impute_method = impute_method)


  # rearrange the columns
  column_order = ['country', 'year', 'region', 'population', 'population_imputed',
                  'total_deceased_donors', 'total_deceased_donors_imputed'] + list(organs_clean.columns)
  column_order = pd.unique(column_order)
  organs_clean = organs_clean.reindex(columns=column_order)

  return organs_clean



### Checking the cleaned and pre-processed data

Let's do some exploration to ensure that the data was cleaned and pre-processed as expected.


In [ ]:
# create the organs_clean object
organs_clean = prepare_organ_data(organs_original)

<ipython-input-75-148f79412c63>:70: FutureWarning:

unique with argument that is not not a Series, Index, ExtensionArray, or np.ndarray is deprecated and will raise in a future version.



Note the `TOTAL Actual DD` variable is now called `total_deceased_donors`, and we decided that the cleaned/pre-processed data should contain *both* an imputed version of this variable *in addition to* the original unimputed version (we would not want this if we were doing predictive modeling, say, but it is fine for exploratory projects such as this one).

Here are the first 10 rows:


In [ ]:
organs_clean.head(10)

country  year  region  population  population_imputed  \
0  Andorra  2007  Europe    100000.0            100000.0   
1  Andorra  2008  Europe    100000.0            100000.0   
2  Andorra  2009  Europe    100000.0            100000.0   
3  Andorra  2010  Europe    100000.0            100000.0   
4  Andorra  2011  Europe    100000.0            100000.0   
5  Andorra  2012  Europe    100000.0            100000.0   
6  Andorra  2013  Europe    100000.0            100000.0   
7  Andorra  2014  Europe    100000.0            100000.0   
8  Andorra  2015  Europe         NaN            100000.0   
9  Andorra  2016  Europe         NaN            100000.0   

   total_deceased_donors  total_deceased_donors_imputed  \
0                    NaN                            0.0   
1                    NaN                            0.0   
2                    NaN                            0.0   
3                    NaN                            0.0   
4                    NaN                            0.0   
5                    NaN                            0.0   
6                    NaN                            0.0   
7                    NaN                            0.0   
8                    NaN                            0.0   
9                    NaN                            0.0   

   deceased_donors_brain_death  deceased_donors_circulatory_death  \
0                          NaN                                NaN   
1                          NaN                                NaN   
2                          NaN                                NaN   
3                          NaN                                NaN   
4                          NaN                                NaN   
5                          NaN                                NaN   
6                          NaN                                NaN   
7                          NaN                                NaN   
8                          NaN                                NaN   
9                          NaN                                NaN   

   total_utilized_deceased_donors  utilized_deceased_donors_brain_death  \
0                             NaN                                   NaN   
1                             NaN                                   NaN   
2                             NaN                                   NaN   
3                             NaN                                   NaN   
4                             NaN                                   NaN   
5                             NaN                                   NaN   
6                             NaN                                   NaN   
7                             NaN                                   NaN   
8                             NaN                                   NaN   
9                             NaN                                   NaN   

   utilized_deceased_donors_circulatory_death  deceased_kidney_tx  \
0                                         NaN                 NaN   
1                                         NaN                 NaN   
2                                         NaN                 NaN   
3                                         NaN                 NaN   
4                                         NaN                 NaN   
5                                         NaN                 NaN   
6                                         NaN                 NaN   
7                                         NaN                 NaN   
8                                         NaN                 NaN   
9                                         NaN                 NaN   

   living_kidney_tx  total_kidney_tx  deceased_liver_tx  domino_liver_tx  \
0               NaN              NaN                NaN              NaN   
1               NaN              NaN                NaN              NaN   
2               NaN              NaN                NaN              NaN   
3               NaN              NaN      

and a random set of 10 rows:

In [ ]:
organs_clean.sample(10)

country  year           region  population  \
1916                 Latvia  2008           Europe   2300000.0   
226                Barbados  2010          America    300000.0   
2493       Papua New Guinea  2009  Western Pacific   6700000.0   
1690  Saint Kitts and Nevis  2016          America         NaN   
1100                  Gabon  2002           Africa   1300000.0   
1254                 Greece  2012           Europe  11400000.0   
2499       Papua New Guinea  2015  Western Pacific         NaN   
3484               Zimbabwe  2010           Africa  12600000.0   
2108             Mauritania  2002           Africa   2800000.0   
114                  Angola  2006           Africa  16400000.0   

      population_imputed  total_deceased_donors  \
1916           2300000.0                   30.0   
226             300000.0                    NaN   
2493           6700000.0                    NaN   
1690            100000.0                    NaN   
1100           1300000.0                    NaN   
1254          11400000.0                   77.0   
2499           7500000.0                    NaN   
3484          12600000.0                    NaN   
2108           2800000.0                    NaN   
114           16400000.0                    NaN   

      total_deceased_donors_imputed  deceased_donors_brain_death  \
1916                           30.0                         19.0   
226                             0.0                          NaN   
2493                            0.0                          NaN   
1690                            0.0                          NaN   
1100                            0.0                          NaN   
1254                           77.0                         77.0   
2499                            0.0                          NaN   
3484                            0.0                          NaN   
2108                            0.0                          NaN   
114                             0.0                          NaN   

      deceased_donors_circulatory_death  total_utilized_deceased_donors  \
1916                               11.0                             NaN   
226                                 NaN                             NaN   
2493                                NaN                             NaN   
1690                                NaN                             NaN   
1100                                NaN                             NaN   
1254                                0.0                             NaN   
2499                                NaN                             NaN   
3484                                NaN                             NaN   
2108                                NaN                             NaN   
114                                 NaN                             NaN   

      utilized_deceased_donors_brain_death  \
1916                                   NaN   
226                                    NaN   
2493                                   NaN   
1690                                   NaN   
1100                                   NaN   
1254                                   NaN   
2499                                   NaN   
3484                                   NaN   
2108                                   NaN   
114                                    NaN   

      utilized_deceased_donors_circulatory_death  deceased_kidney_tx  \
1916                                         NaN                53.0   
226                                          NaN                 NaN   
2493                                         NaN                 NaN   
1690                                         NaN                 NaN   
1100                                         NaN                 NaN   
1254                                         NaN               130.0   
2499                                         NaN                 NaN   
3484                                         NaN                 NaN   
2

The table below shows the relevant columns for the cleaned data for Peru. Note that the `TOTAL Actual DD` variable is now called `total_deceased_donors`, and the imputed version is called `total_deceased_donors_imputed`.

In [ ]:
organs_clean.query('country == "Peru"')[["country", "year", "population", "total_deceased_donors", "total_deceased_donors_imputed"]]

country  year  population  total_deceased_donors  \
2466    Peru  2000  25700000.0                    NaN   
2467    Peru  2001  26100000.0                    NaN   
2468    Peru  2002  26500000.0                    NaN   
2469    Peru  2003  27200000.0                    NaN   
2470    Peru  2004  27600000.0                   29.0   
2471    Peru  2005  28000000.0                   22.0   
2472    Peru  2006  28400000.0                    NaN   
2473    Peru  2007  28800000.0                    NaN   
2474    Peru  2008  28200000.0                    NaN   
2475    Peru  2009  29200000.0                    NaN   
2476    Peru  2010  29500000.0                   94.0   
2477    Peru  2011  29400000.0                  127.0   
2478    Peru  2012  29700000.0                   94.0   
2479    Peru  2013  30400000.0                   97.0   
2480    Peru  2014  30800000.0                   73.0   
2481    Peru  2015  31400000.0                   82.0   
2482    Peru  2016  31800000.0                   70.0   
2483    Peru  2017  32200000.0                   52.0   

      total_deceased_donors_imputed  
2466                           29.0  
2467                           29.0  
2468                           29.0  
2469                           29.0  
2470                           29.0  
2471                           22.0  
2472                           58.0  
2473                           58.0  
2474                           58.0  
2475                           58.0  
2476                           94.0  
2477                          127.0  
2478                           94.0  
2479                           97.0  
2480                           73.0  
2481                           82.0  
2482                           70.0  
2483                           52.0

**Let's reconvene to start exploratory data analysis**

##[DSLC stage 2]: Exploratory Data Analysis


To do so we will ask a question about the data and then produce several exploratory visualizations to answer the following question:

*Are global organ donations are increasing over time?*



In [ ]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots

Since many of our explorations will involve looking at the donor rates, let's create a version of the original and imputed donor counts per million. How would you complete this *pre-processing featurization* step? (Would it be better to include in  prepareOrganData()?)

In [ ]:
# add a donors_per_mil column for the donors cols
organs_clean = organs_clean.assign(total_deceased_donors_per_mil = (organs_clean.total_deceased_donors / (organs_clean.population_imputed + 1)) * 1_000_000)
organs_clean = organs_clean.assign(total_deceased_donors_imputed_per_mil = (organs_clean.total_deceased_donors_imputed / (organs_clean.population_imputed + 1)) * 1_000_000)
  # note that we use `population_imputed + 1` in the denominator because there are some countries with a reported population of 0.

## Are global organ donations increasing over time?

Create a plot that uses imputed organ donations across the world over time to answer this question. The imputed donor counts are based on the "average" imputation method.  


In [ ]:
donors_by_year = organs_clean.groupby("year")["total_deceased_donors_imputed"].sum()
px.line(donors_by_year)

## What country has the highest number of donors?

What data can you use to answer this question and how can you visualize it?



In [ ]:
organs_2017 = organs_clean[organs_clean["year"] == 2017]
countries_top_20_2017 = organs_2017.set_index("country")["total_deceased_donors_imputed"] \
    .sort_values(ascending=False) \
    .head(20)
countries_top_20_2017

In [ ]:
px.bar(countries_top_20_2017)

## What country has the highest donation rate?

For the countries with the top 20 highest donation rates in 2017, create a heatmap to compare their donation rates between 2007-2017

In [ ]:
countries_top_20_2017_per_mil = organs_2017.set_index("country")["total_deceased_donors_imputed_per_mil"] \
    .sort_values(ascending=False) \
    .head(20)
countries_top_20_2017_per_mil



In [ ]:
# extract the names of the top 20 countries from 2017
countries_top_2017 = countries_top_20_2017_per_mil.index
# filter the organs_clean data (all years) to these top 20 countries
organs_top_countries = organs_clean.query('country in @countries_top_2017').copy()
# add the word "year" to the year column so that we can use it as a variable name later
organs_top_countries["year"] = "year_" + organs_top_countries["year"].astype(str)
# select just the three columns: year, country, total_deceased_donors_imputed_per_mil
organs_top_countries_per_mil = organs_top_countries[["year", "country", "total_deceased_donors_imputed_per_mil"]]
# spread the data across year
organs_top_countries_per_mil_wide = (organs_top_countries_per_mil.pivot(
    index="country",
    columns="year",
    values="total_deceased_donors_imputed_per_mil"
  ).sort_values("year_2017", ascending=False))
px.imshow(organs_top_countries_per_mil_wide, color_continuous_scale="Greys")

Another way to visualize the top donor rates over time is by using overlaying line plots.

In [ ]:
# Top 20 countries in 2017
# filter to the top 20 countries in 2017
organs_clean_2017 = organs_clean.query('year == 2017').copy()
top_20_countries = organs_clean_2017.nlargest(n=20, columns="total_deceased_donors_imputed_per_mil")["country"]
organs_clean_top_20 = organs_clean.query('country in @top_20_countries').copy()
# add highlight for spain
organs_clean_top_20["spain"] = organs_clean_top_20["country"] == "Spain"
# add highlight for Spain, USA, and Croatia
organs_clean_top_20["highlight"] = np.where(organs_clean_top_20["country"].isin(["Spain", "United States of America", "Croatia"]),
                                            organs_clean_top_20["country"],
                                            "Other")

fig= px.line(organs_clean_top_20,
                  x="year",
                  y="total_deceased_donors_per_mil",
                  color="highlight",
                  color_discrete_sequence=["grey", "#84ACCE", "#F6AE2D", "#589D6F"],
                  hover_name="country",
                  hover_data={"year": True,
                              "total_deceased_donors_per_mil": ":.2f",
                              "highlight": False})
fig.update_traces(opacity=1,
                  line=dict(width=5))

fig.update_layout(yaxis_title="Organ donations per million",
                      legend_title="Country",
                      plot_bgcolor="rgba(0,0,0,0)",
                      showlegend=False)
# add direct country annotation
fig.add_annotation(x=2017, y=47,
            text="Spain",
            showarrow=False, xanchor="left")
fig.add_annotation(x=2017, y=34,
            text="Croatia",
            showarrow=False, xanchor="left")
fig.add_annotation(x=2017, y=31,
            text="USA",
            showarrow=False, xanchor="left")
# change the specs for the "Other" line only
fig['data'][0]['line']['width'] = 1
fig['data'][0]['opacity'] = 0.25
fig.show()

## Do countries with larger populations typically have more donors?

This may be a reasonable assumption, but how can we check this in our data?

In [ ]:
# compute correlation between imputed population and number of donors
organs_clean_2017["population_imputed"].corr(organs_clean_2017["total_deceased_donors_imputed"])

In [ ]:
fig = px.scatter(organs_clean_2017,
                 x="population_imputed",
                 y="total_deceased_donors_imputed",
                 hover_name="country")
fig.add_annotation(x=1_400_000_000, y=4_100,
            text="China",
            showarrow=False, xanchor="left")
fig.add_annotation(x=340_000_000, y=10_300,
            text="USA",
            showarrow=False, xanchor="left")
fig.add_annotation(x=220_000_000, y=3_450,
            text="Brazil",
            showarrow=False, xanchor="left")
fig.show()

How can we fix this plot to see if there is a relationship? Can you remove outliers or change scale to do so?

*Hint: consider log scale*

In [ ]:
#TODO: Remove outliers and plot data again

In [ ]:
#TODO: Plot the data without outliers again on log scale

## More for you to explore
1. Can you use scatter plots to compare donor rates for kidneys, livers, hearts, and lungs between the United States and Spain?
2. Would it be helpful for you to do more cleaning of the data such as dropping columns? Why and how would this be done?
3. What other questions can you ask about this data? Would you need to get additional data to explore your questions and if so, what data?